### PR TITLE
Bug/6593/fix send email for user in patch update status

### PR DESCRIPTION
**Pull request by VasylyshynDmytro**
[Link on task from board](https://github.com/orgs/ita-social-projects/projects/43/views/4?pane=issue&itemId=69330459&issue=ita-social-projects%7CGreenCity%7C6593)
image
image

add logiс for update place status in db(service), wright tests, refactor old update status and create a new DTO 

_(to be filled out by PR submitter)_
- [ ] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

**Comment by Cr1stal423** [comment]
Fix issue

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 1]
```diff
@@ -2,22 +2,12 @@
...
-import greencity.dto.place.FilterPlaceCategory;
-import greencity.dto.place.PlaceAddDto;
-import greencity.dto.place.PlaceByBoundsDto;
-import greencity.dto.place.PlaceInfoDto;
-import greencity.dto.place.PlaceResponse;
-import greencity.dto.place.PlaceUpdateDto;
-import greencity.dto.place.PlaceVO;
-import greencity.dto.place.PlaceWithUserDto;
-import greencity.dto.place.UpdatePlaceStatusDto;
+import greencity.dto.place.*;
```
Comment (Cr1stal423) [inline core/src/main/java/greencity/controller/PlaceController.java]: wildcard imports

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 2]
```diff
@@ -1,6 +1,11 @@
 package greencity.controller;
 
+import greencity.constant.ErrorMessage;
 import greencity.converters.UserArgumentResolver;
+import greencity.dto.place.*;
```
Comment (Cr1stal423) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: wildcard import

File: service-api/src/main/java/greencity/message/ScheduledEmailMessage.java [hunk 3]
```diff
@@ -1,14 +1,12 @@
 package greencity.message;
 
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
Comment (Cr1stal423) [inline service-api/src/main/java/greencity/message/ScheduledEmailMessage.java]: wildcard import

File: service-api/src/main/java/greencity/service/PlaceService.java [hunk 4]
```diff
@@ -2,18 +2,7 @@
...
-import greencity.dto.place.FilterAdminPlaceDto;
-import greencity.dto.place.FilterPlaceCategory;
-import greencity.dto.place.PlaceAddDto;
-import greencity.dto.place.PlaceByBoundsDto;
-import greencity.dto.place.PlaceInfoDto;
-import greencity.dto.place.PlaceResponse;
-import greencity.dto.place.PlaceUpdateDto;
-import greencity.dto.place.PlaceVO;
-import greencity.dto.place.UpdatePlaceStatusDto;
+import greencity.dto.place.*;
```
Comment (Cr1stal423) [inline service-api/src/main/java/greencity/service/PlaceService.java]: wildcard import

File: service/src/test/java/greencity/service/PlaceServiceImplTest.java [hunk 5]
```diff
@@ -9,18 +9,7 @@
...
-import greencity.dto.place.FilterAdminPlaceDto;
-import greencity.dto.place.FilterPlaceCategory;
-import greencity.dto.place.PlaceAddDto;
-import greencity.dto.place.PlaceByBoundsDto;
-import greencity.dto.place.PlaceInfoDto;
-import greencity.dto.place.PlaceResponse;
-import greencity.dto.place.PlaceUpdateDto;
-import greencity.dto.place.PlaceVO;
-import greencity.dto.place.UpdatePlaceStatusDto;
+import greencity.dto.place.*;
```
Comment (Cr1stal423) [inline service/src/test/java/greencity/service/PlaceServiceImplTest.java]: wildcard import

File: service/src/main/java/greencity/service/PlaceServiceImpl.java [hunk 6]
```diff
@@ -13,18 +13,7 @@
...
-import greencity.dto.place.FilterAdminPlaceDto;
-import greencity.dto.place.FilterPlaceCategory;
-import greencity.dto.place.PlaceAddDto;
-import greencity.dto.place.PlaceByBoundsDto;
-import greencity.dto.place.PlaceInfoDto;
-import greencity.dto.place.PlaceResponse;
-import greencity.dto.place.PlaceUpdateDto;
-import greencity.dto.place.PlaceVO;
-import greencity.dto.place.UpdatePlaceStatusDto;
+import greencity.dto.place.*;
```
Comment (Cr1stal423) [inline service/src/main/java/greencity/service/PlaceServiceImpl.java]: wildcard imports

File: service-api/src/test/java/greencity/client/RestClientTest.java [hunk 7]
```diff
@@ -55,11 +57,8 @@
 import org.springframework.web.context.request.RequestContextHolder;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
-import static org.mockito.Mockito.any;
-import static org.mockito.Mockito.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.*;
```
Comment (Cr1stal423) [inline service-api/src/test/java/greencity/client/RestClientTest.java]: wildcard imports

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 8]
```diff
@@ -55,18 +53,16 @@
 import static greencity.ModelUtils.getUserVO;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (Cr1stal423) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: wildcard import

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 9]
```diff
@@ -55,18 +53,16 @@
...
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 import static greencity.ModelUtils.getPrincipal;
 import static greencity.enums.PlaceStatus.APPROVED;
 import static greencity.enums.PlaceStatus.PROPOSED;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
```
Comment (Cr1stal423) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: wildcard import

File: service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+@ToString
+public class UpdatePlaceStatusWithUserEmailDto {
+    @NotNull
+    private String placeName;
+
+    @NotNull
+    private PlaceStatus newStatus;
+
+    @NotNull
+    private String userName;
```
Comment (Cr1stal423) [inline service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java:22]: For String, it is better to use @NotBlank

File: core/src/main/java/greencity/controller/PlaceController.java (hunk 1 above)
Comment (holotsvan) [inline core/src/main/java/greencity/controller/PlaceController.java]: we do not use wild card imports

File: core/src/test/java/greencity/controller/PlaceControllerTest.java (hunk 2 above)
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: wildcard

File: core/src/test/java/greencity/controller/PlaceControllerTest.java (hunk 8 above)
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: wildcard

File: service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package greencity.dto.place;
+
+import greencity.enums.PlaceStatus;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
```
Comment (holotsvan) [inline service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java]: Could use Data annotation which combines Getter Setter and ToString

File: service-api/src/main/java/greencity/message/ScheduledEmailMessage.java [hunk 12]
```diff
@@ -1,14 +1,12 @@
...
 
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
 @Builder
+@Getter
```
Comment (holotsvan) [inline service-api/src/main/java/greencity/message/ScheduledEmailMessage.java]: Data already contains equivalent to Getter in itself

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 13]
```diff
@@ -606,4 +580,126 @@ void getAllPlaces() throws Exception {
         verify(placeService, times(1))
             .findAll(pageable, principal);
     }
+
+    @Test
+    void updateStatusWithApprovedStatusAPPROVED() throws Exception {
+        UpdatePlaceStatusWithUserEmailDto dto = new UpdatePlaceStatusWithUserEmailDto();
+        dto.setPlaceName("testPlace");
```
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: mb, move creation of this dto to model utils, because it is used in 4 tests and the only difference is newStatus

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 14]
```diff
@@ -606,4 +580,126 @@ void getAllPlaces() throws Exception {
...
+
+        mockMvc.perform(patch(placeLink + "/status")
+            .contentType(MediaType.APPLICATION_JSON)
+            .content(json))
+            .andExpect(status().isOk())
+            .andExpect(content().string("Status updated successfully for place: testPlace"));
+        verify(placeService, times(1)).updatePlaceStatus(any(UpdatePlaceStatusWithUserEmailDto.class));
+        verify(restClient, times(0))
+            .sendEmailNotificationChangesPlaceStatus(any(UpdatePlaceStatusWithUserEmailDto.class));
+    }
```
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java:626]: 4 tests looks similar except status and that APPROVED, DECLINED send email notification about changes places status and PROPOSED, DELTED don't. mb, would be better to use parameterized tests, imho

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 15]
```diff
@@ -606,4 +580,126 @@ void getAllPlaces() throws Exception {
...
+
+    @Test
+    void updateStatusDoesNotSendNotificationButSavesDataWhenStatusIsDELETED() throws Exception {
+        UpdatePlaceStatusWithUserEmailDto dto = new UpdatePlaceStatusWithUserEmailDto();
+        dto.setPlaceName("testPlace");
+        dto.setNewStatus(PlaceStatus.DELETED);
+        dto.setUserName("testUser");
+        dto.setEmail("user@example.com");
+        Place place = new Place();
```
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: Could not see where place is used in this test. Are we realy need it there?

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 16]
```diff
@@ -606,4 +580,126 @@ void getAllPlaces() throws Exception {
...
+    void updateStatusDoesNotSendNotificationButSavesDataWhenStatusIsPROPOSED() throws Exception {
+        UpdatePlaceStatusWithUserEmailDto dto = new UpdatePlaceStatusWithUserEmailDto();
+        dto.setPlaceName("testPlace");
+        dto.setNewStatus(PlaceStatus.PROPOSED);
+        dto.setUserName("testUser");
+        dto.setEmail("user@example.com");
+        Place place = new Place();
+        place.setId(1L);
+        place.setName("testPlace");
+        place.setStatus(PlaceStatus.PROPOSED);
```
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: Could not see where place is used in this test. Are we really need it there?

File: service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
+package greencity.dto.place;
+
+import greencity.enums.PlaceStatus;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
```
Comment (KizerovDmitriy) [inline service-api/src/main/java/greencity/dto/place/UpdatePlaceStatusWithUserEmailDto.java]: we don't use wildcard imports

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 18]
```diff
@@ -300,12 +303,15 @@ public ResponseEntity<List<PlaceByBoundsDto>> getFilteredPlaces(
...
         @ApiResponse(responseCode = "403", description = HttpStatuses.FORBIDDEN,
-            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN)))
+            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN))),
     })
     @PatchMapping("/status")
-    public ResponseEntity<UpdatePlaceStatusDto> updateStatus(@Valid @RequestBody UpdatePlaceStatusDto dto) {
-        return ResponseEntity.status(HttpStatus.OK)
-            .body(placeService.updateStatus(dto.getId(), dto.getStatus()));
+    public ResponseEntity<String> updateStatus(@Valid @RequestBody UpdatePlaceStatusWithUserEmailDto dto) {
+        if (dto.getNewStatus() == PlaceStatus.APPROVED || dto.getNewStatus() == PlaceStatus.DECLINED) {
```
Comment (KizerovDmitriy) [inline core/src/main/java/greencity/controller/PlaceController.java]: Maybe better move this logic to service layer?

**Comment by KizerovDmitriy** [comment]
Resolve this 5 issues:
image

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 19]
```diff
@@ -300,12 +302,16 @@ public ResponseEntity<List<PlaceByBoundsDto>> getFilteredPlaces(
...
         @ApiResponse(responseCode = "403", description = HttpStatuses.FORBIDDEN,
-            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN)))
+            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN))),
     })
     @PatchMapping("/status")
-    public ResponseEntity<UpdatePlaceStatusDto> updateStatus(@Valid @RequestBody UpdatePlaceStatusDto dto) {
-        return ResponseEntity.status(HttpStatus.OK)
-            .body(placeService.updateStatus(dto.getId(), dto.getStatus()));
+    public ResponseEntity<String> updateStatus(@Valid @RequestBody UpdatePlaceStatusWithUserEmailDto dto) {
+        try {
```
Comment (KizerovDmitriy) [inline core/src/main/java/greencity/controller/PlaceController.java]: You don't need to wrap service call method in try/catch block. If exception will be throwed in service layer, the [global exception handler](https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc) was catch this exception and return correct status code and message.

**Review by KizerovDmitriy — CHANGES_REQUESTED** [review]

**Comment by KizerovDmitriy** [comment]
Increase your code coverage to up the 85% or higher.
image

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 20]
```diff
@@ -300,12 +301,12 @@ public ResponseEntity<List<PlaceByBoundsDto>> getFilteredPlaces(
...
             content = @Content(examples = @ExampleObject(HttpStatuses.UNAUTHORIZED))),
         @ApiResponse(responseCode = "403", description = HttpStatuses.FORBIDDEN,
-            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN)))
+            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN))),
     })
     @PatchMapping("/status")
-    public ResponseEntity<UpdatePlaceStatusDto> updateStatus(@Valid @RequestBody UpdatePlaceStatusDto dto) {
-        return ResponseEntity.status(HttpStatus.OK)
-            .body(placeService.updateStatus(dto.getId(), dto.getStatus()));
+    public ResponseEntity<String> updateStatus(@Valid @RequestBody UpdatePlaceStatusWithUserEmailDto dto) {
```
Comment (KizerovDmitriy) [inline core/src/main/java/greencity/controller/PlaceController.java]: You need to return UpdatePlaceStatusDto, not String.

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 21]
```diff
@@ -300,12 +301,12 @@ public ResponseEntity<List<PlaceByBoundsDto>> getFilteredPlaces(
...
+        placeService.updatePlaceStatus(dto);
+        return ResponseEntity.ok("Status updated successfully for place: " + dto.getPlaceName());
```
Comment (KizerovDmitriy) [inline core/src/main/java/greencity/controller/PlaceController.java]: Replace this, to this: return ResponseEntity.ok(placeService.updateStatus(dto));

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 22]
```diff
@@ -56,17 +57,17 @@
...
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 import static greencity.ModelUtils.getPrincipal;
 import static greencity.enums.PlaceStatus.APPROVED;
 import static greencity.enums.PlaceStatus.PROPOSED;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
```
Comment (holotsvan) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java]: still wildcard

**Review by holotsvan — APPROVED** [review]

**Review by KizerovDmitriy — APPROVED** [review]

File: service/src/main/java/greencity/service/PlaceServiceImpl.java [hunk 23]
```diff
@@ -635,4 +638,31 @@ private PageableDto<SearchPlacesDto> getSearchPlacesDtoPageableDto(Page<Place> p
...
+     *            the new status.
+     * @return The updated UpdatePlaceStatusWithUserEmailDto.
+     * @throws NotFoundException If the place or user is not found.
+     */
+    @Override
+    public UpdatePlaceStatusWithUserEmailDto updatePlaceStatus(UpdatePlaceStatusWithUserEmailDto dto) {
+        Place place = placeRepo.findByNameIgnoreCase(dto.getPlaceName())
+            .orElseThrow(() -> new NotFoundException(ErrorMessage.PLACE_NOT_FOUND_BY_NAME + dto.getPlaceName()));
+
+        if (userRepo.findByEmail(dto.getEmail()).isEmpty()) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/PlaceServiceImpl.java:656]: please move this check before `placeRepo.findByNameIgnoreCase`
you call `placeRepo.findByNameIgnoreCase` event if a user is not found

File: service-api/src/main/java/greencity/client/RestClient.java [hunk 24]
```diff
@@ -427,6 +428,14 @@ public void sendInterestingEcoNews(InterestingEcoNewsDto message) {
             + RestTemplateLinks.SEND_INTERESTING_ECO_NEWS, HttpMethod.POST, entity, Object.class);
     }
 
+    public void sendEmailNotificationChangesPlaceStatus(UpdatePlaceStatusWithUserEmailDto message) {
+        HttpHeaders headers = setHeader();
+        headers.setContentType(MediaType.APPLICATION_JSON);
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/client/RestClient.java:433]: heder can be a constant

File: dao/src/main/java/greencity/repository/PlaceRepo.java [hunk 25]
```diff
@@ -79,4 +80,13 @@ List<Place> findAllByModifiedDateBetweenAndStatus(ZonedDateTime startDate, Zoned
...
     List<Place> findPlaceByCategory(String[] category);
+
+    /**
+     * Finds a place by its name.
+     *
+     * @param name the name of the place.
+     * @return an optional containing the place if found, or empty otherwise.
+     */
+    @Query("SELECT p FROM Place p WHERE LOWER(p.name) = LOWER(:name)")
+    Optional<Place> findByNameIgnoreCase(@Param("name") String name);
```
Comment (RomaMocherniuk) [inline dao/src/main/java/greencity/repository/PlaceRepo.java:91]: is the name the unique field?
if not we will have a list here

File: core/src/test/java/greencity/controller/PlaceControllerTest.java [hunk 26]
```diff
@@ -428,28 +429,6 @@ void getFilteredPlaces() throws Exception {
         verify(placeService).getPlacesByFilter(filterPlaceDto, userVO);
     }
 
-    @Test
-    void updateStatus() throws Exception {
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/PlaceControllerTest.java:432]: why did you remove this test?

File: core/src/main/java/greencity/controller/PlaceController.java [hunk 27]
```diff
@@ -300,12 +301,12 @@ public ResponseEntity<List<PlaceByBoundsDto>> getFilteredPlaces(
         @ApiResponse(responseCode = "401", description = HttpStatuses.UNAUTHORIZED,
             content = @Content(examples = @ExampleObject(HttpStatuses.UNAUTHORIZED))),
         @ApiResponse(responseCode = "403", description = HttpStatuses.FORBIDDEN,
-            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN)))
+            content = @Content(examples = @ExampleObject(HttpStatuses.FORBIDDEN))),
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/PlaceController.java]: redundant `,`

File: core/src/test/java/greencity/controller/PlaceControllerTest.java (hunk 26 above)
Comment (VasylyshynDmytro) [reply core/src/test/java/greencity/controller/PlaceControllerTest.java:432]: The task involved fixing the old code, so I modified the existing method to ensure it sends an email to the user and saves valid data in the database. Since the test was written for the method I modified, I deleted the old test and added new tests methods to cover the updated functionality.

File: dao/src/main/java/greencity/repository/PlaceRepo.java (hunk 25 above)
Comment (VasylyshynDmytro) [reply dao/src/main/java/greencity/repository/PlaceRepo.java:91]: In general, we don't have any annotations on the placeName field to guarantee uniqueness, nor is there a uniqueness check in the service's creation method. However, the task was described as if the name of each place is unique. The task also specified that we accept the following JSON structure:
{
  "placeName": "",
  "newStatus": "",
  "userName": "",
  "email": ""
}
Since there’s no unique field to rely on, I implemented it this way.

File: service/src/main/java/greencity/service/PlaceServiceImpl.java (hunk 23 above)
Comment (VasylyshynDmytro) [reply service/src/main/java/greencity/service/PlaceServiceImpl.java:656]: I was guided by the logic that there are fewer places than users, so it’s better to first check if such places exist, and only then check if such users exist. Or does it not matter much, and should non-existent users be filtered out first?

**Review by RomaMocherniuk — APPROVED** [review]

**Review by Cr1stal423 — APPROVED** [review]
